packages/@aws-cdk/aws-codepipeline/lib/cross-region-scaffold-stack.ts

```diff
@@ -0,0 +1,61 @@
+import s3 = require('@aws-cdk/aws-s3');
+import cdk = require('@aws-cdk/cdk');
+import crypto = require('crypto');
+
+/**
+ * Construction properties for {@link CrossRegionScaffoldStack}.
+ */
+export interface CrossRegionScaffoldStackProps {
+  region: string;
+
+  account: string;
+}
+
+/**
+ * A Stack containing resources required for the cross-region CodePipeline functionality to work.
+ */
+export class CrossRegionScaffoldStack extends cdk.Stack {
+  /**
+   * The name of the S3 Bucket used for replicating the Pipeline's artifacts into the region.
+   */
+  public readonly replicationBucket: string;
+
+  constructor(parent?: cdk.App, props?: CrossRegionScaffoldStackProps) {
+    super(parent, generateStackName(props || defaultCrossRegionScaffoldStackProps()), {
+      env: {
+        region: (props || defaultCrossRegionScaffoldStackProps()).region,
+        account: (props || defaultCrossRegionScaffoldStackProps()).account,
+      },
+    });
+
+    props = props || defaultCrossRegionScaffoldStackProps();
+
+    const replicationBucketName = generateUniqueName('cdk-cross-region-codepipeline-replication-bucket-',
+      props.region, props.account, false, 12);
+    new s3.Bucket(this, 'CrossRegionCodePipelineReplicationBucket', {
+      bucketName: replicationBucketName,
+    });
+    this.replicationBucket = replicationBucketName;
+  }
+}
+
+function generateStackName(props: CrossRegionScaffoldStackProps): string {
+  return `aws-cdk-codepipeline-cross-region-scaffolding-${props.region}`;
+}
+
+function generateUniqueName(baseName: string, region: string, account: string,
+                            toUpperCase: boolean, hashPartLen: number = 8): string {
+  const sha256 = crypto.createHash('sha256')
+    .update(baseName)
+    .update(region)
+    .update(account);
+
+  const hash = sha256.digest('hex').slice(0, hashPartLen);
+
+  return baseName + (toUpperCase ? hash.toUpperCase() : hash.toLowerCase());
+}
+
+// purely to defeat the limitation that a required argument cannot follow an optional one
```

Comment:
So why make the `parent` of `CrossRegionScaffoldStack` optional?

Reply:
Because `parent` is optional for `Stack`.

Reply:
It can't be optional in `CrossRegionScaffoldStack`, because the parent is taken from the `Stack` the Pipeline itself is in. If `parent` is optional there, it needs to be optional here as well. See code: https://github.com/awslabs/aws-cdk/blob/master/packages/%40aws-cdk/aws-codepipeline/lib/pipeline.ts#L308